README.md
# bitset-word8

[![License: MIT](https://img.shields.io/badge/License-MIT-brightgreen.svg)](https://opensource.org/licenses/MIT)
[![Build Status](https://travis-ci.org/nshimaza/bitset-word8.svg?branch=master)](https://travis-ci.org/nshimaza/bitset-word8)
[![Hackage](https://img.shields.io/hackage/v/bitset-word8.svg?style=flat)](https://hackage.haskell.org/package/bitset-word8)
[![Stackage Nightly](http://stackage.org/package/bitset-word8/badge/nightly)](http://stackage.org/nightly/package/bitset-word8)
[![Stackage LTS](http://stackage.org/package/bitset-word8/badge/lts)](http://stackage.org/lts/package/bitset-word8)

Space efficient set of `Word8` and some pre-canned sets useful for parsing HTTP related `ByteString`.
This package is intended to provide O(1) membership test on any subset of ASCII and Latin-1 character set
in order to write efficient HTTP related parser.

### Creating your own set

You can create your own set by `fromList`.

```haskell
myCharSet :: BitSetWord8
myCharSet = fromList [ 'Y', 'y', 'N', 'n' ]
```

You can create pre-evaluated set using Template Haskell.

```haskell
{-# LANGUAGE TemplateHaskell #-}
import Language.Haskell.TH.Syntax (Lift, lift)

myPreEvaluatedCharSet :: BitSetWord8
myPreEvaluatedCharSet = $(lift myCharSet)
```

### Example Usage

```haskell
import Data.Attoparsec.ByteString

-- | Parse RFC7230 token.
token :: Parser ByteString
token = takeWhile1 (member rfc7230TChar)
```

CHANGELOG.md
## Bitset-word8 0.1.1.0

* Changed backend from ByteString to unboxed Word64 based on benchmark result.
* Added benchmarks with some backend implementations.

## Bitset-word8 0.1.0.1

* Minor modifications in documents.

## Bitset-word8 0.1.0.0

* Initial release.


test/Data/BitSetWord8Spec.hs
module Data.BitSetWord8Spec where

import           Data.Char             (chr)
import           Data.Word             (Word8)

import           Test.Hspec
import           Test.Hspec.QuickCheck (prop)
import           Test.QuickCheck

import           Data.BitSetWord8

spec :: Spec
spec = do
    describe "fromList" $ do
        prop "creates empty BitSetWord8 from empty list" $ forAll (choose (0, 0xff :: Word8)) $ \n -> do
            member (fromList []) n `shouldBe` False

        prop "creates fullfilled BitSetWord8 from fullfilled list" $ forAll (choose (0, 0xff :: Word8)) $ \n -> do
            member (fromList [chr 0 .. chr 0xff]) n `shouldBe` True

    describe "rfc5234Digit'" $ do
        prop "recognizes any digit" $ forAll (choose (0x30, 0x39 :: Word8)) $ \n -> do
            member (fromList rfc5234Digit') n `shouldBe` True

        prop "unrecognizes any non-digit characters (lower)" $ forAll (choose (0, 0x2f :: Word8)) $ \n -> do
            member (fromList rfc5234Digit') n `shouldBe` False

        prop "unrecognizes any non-digit characters (higher)" $ forAll (choose (0x3a, 0xff :: Word8)) $ \n -> do
            member (fromList rfc5234Digit') n `shouldBe` False

    describe "rfc5234Alpha'" $ do
        prop "recognizes any upper case alphabet" $ forAll (choose (0x41, 0x5a :: Word8)) $ \n -> do
            member (fromList rfc5234Alpha') n `shouldBe` True

        prop "recognizes any lower case alphabet" $ forAll (choose (0x61, 0x7a :: Word8)) $ \n -> do
            member (fromList rfc5234Alpha') n `shouldBe` True

        prop "unrecognizes any non-alphabet (lower)" $ forAll (choose (0, 0x40 :: Word8)) $ \n -> do
            member (fromList rfc5234Alpha') n `shouldBe` False

        prop "unrecognizes any non-alphabet (middle)" $ forAll (choose (0x5b, 0x60 :: Word8)) $ \n -> do
            member (fromList rfc5234Alpha') n `shouldBe` False

        prop "unrecognizes any non-alphabet (higher)" $ forAll (choose (0x7b, 0xff :: Word8)) $ \n -> do
            member (fromList rfc5234Alpha') n `shouldBe` False



bench/Bench.hs
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DeriveGeneric  #-}
{-# LANGUAGE MagicHash      #-}
{-# LANGUAGE UnboxedTuples  #-}

import           Control.DeepSeq     (NFData, force)
import           Criterion.Main
import           Data.Bits           (testBit)
import           Data.ByteString     (ByteString, index, pack)
import           Data.Vector.Unboxed (Vector (..), (!), fromList)
import           Data.Word           (Word64, Word8)
import           GHC.Generics        (Generic)

{-
    Unboxed Word64 backend
-}
data BitSetWord8Word64 = BitSetWord8Word64 {-# UNPACK #-} !Word64
                                           {-# UNPACK #-} !Word64
                                           {-# UNPACK #-} !Word64
                                           {-# UNPACK #-} !Word64
    deriving (Eq, Generic, NFData, Show)

memberWord64 :: BitSetWord8Word64 -> Word8 -> Bool
memberWord64 bitSet val = doMember bitSet (val `div` 64) (val `mod` 64)
  where
    doMember :: BitSetWord8Word64 -> Word8 -> Word8 -> Bool
    doMember (BitSetWord8Word64 w _ _ _) 0 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word64 _ w _ _) 1 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word64 _ _ w _) 2 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word64 _ _ _ w) 3 ind = testBit w (fromIntegral ind)

allZeroWord64 :: BitSetWord8Word64
allZeroWord64 = force $ BitSetWord8Word64 0 0 0 0

allOneWord64 :: BitSetWord8Word64
allOneWord64 = force $ BitSetWord8Word64 maxBound maxBound maxBound maxBound

{-
    ByteString backend
-}
newtype BitSetWord8ByteString = BitSetWord8ByteString ByteString deriving (Eq, Generic, NFData, Show)

memberByteString :: BitSetWord8ByteString -> Word8 -> Bool
memberByteString (BitSetWord8ByteString bs) w = testBit (index bs (fromIntegral (w `div` 8))) (fromIntegral (w `mod` 8))

allZeroByteString :: BitSetWord8ByteString
allZeroByteString = force $ BitSetWord8ByteString $ pack [ 0,0,0,0,0,0,0,0
                                                         , 0,0,0,0,0,0,0,0
                                                         , 0,0,0,0,0,0,0,0
                                                         , 0,0,0,0,0,0,0,0
                                                         ]

allOneByteString :: BitSetWord8ByteString
allOneByteString = force $ BitSetWord8ByteString $ pack [ 0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff
                                                        , 0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff
                                                        , 0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff
                                                        , 0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff
                                                        ]

{-
    Vector of Word64 backend
-}
newtype BitSetWord8Vector64 = BitSetWord8Vector64 (Vector Word64) deriving (Eq, Generic, NFData, Show)

memberVector64 :: BitSetWord8Vector64 -> Word8 -> Bool
memberVector64 (BitSetWord8Vector64 v) w = testBit (v ! (fromIntegral (w `div` 64))) (fromIntegral (w `mod` 64))

allZeroVector64 :: BitSetWord8Vector64
allZeroVector64 = force $ BitSetWord8Vector64 $ fromList [ 0,0,0,0 ]

allOneVector64 :: BitSetWord8Vector64
allOneVector64 = force $ BitSetWord8Vector64 $ fromList [ maxBound, maxBound, maxBound, maxBound ]

{-
    Vector of Word8 backend
-}
newtype BitSetWord8Vector8 = BitSetWord8Vector8 (Vector Word8) deriving (Eq, Generic, NFData, Show)

memberVector8 :: BitSetWord8Vector8 -> Word8 -> Bool
memberVector8 (BitSetWord8Vector8 v) w = testBit (v ! (fromIntegral (w `div` 8))) (fromIntegral (w `mod` 8))

allZeroVector8 :: BitSetWord8Vector8
allZeroVector8 = force $ BitSetWord8Vector8 $ fromList [ 0,0,0,0,0,0,0,0
                                                       , 0,0,0,0,0,0,0,0
                                                       , 0,0,0,0,0,0,0,0
                                                       , 0,0,0,0,0,0,0,0
                                                       ]

allOneVector8 :: BitSetWord8Vector8
allOneVector8 = force $ BitSetWord8Vector8 $ fromList [ 0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff
                                                      , 0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff
                                                      , 0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff
                                                      , 0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff
                                                      ]

{-
    Unboxed Word8 backend
-}
data BitSetWord8Word8 = BitSetWord8Word8 {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
                                         {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
                                         {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
                                         {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
                                         {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
                                         {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
                                         {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
                                         {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
                                         {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
                                         {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
                                         {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
                                         {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
                                         {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
                                         {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
                                         {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
                                         {-# UNPACK #-} !Word8 {-# UNPACK #-} !Word8
    deriving (Eq, Generic, NFData, Show)

memberWord8 :: BitSetWord8Word8 -> Word8 -> Bool
memberWord8 bitSet val = doMember bitSet (val `div` 8) (val `mod` 8)
  where
    doMember :: BitSetWord8Word8 -> Word8 -> Word8 -> Bool
    doMember (BitSetWord8Word8 w _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _)  0 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ w _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _)  1 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ w _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _)  2 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ w _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _)  3 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ w _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _)  4 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ w _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _)  5 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ w _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _)  6 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ w  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _)  7 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  w _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _)  8 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ w _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _)  9 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ w _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _) 10 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ w _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _) 11 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ w _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _) 12 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ w _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _) 13 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ w _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _) 14 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ w  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _) 15 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  w _ _ _ _ _ _ _  _ _ _ _ _ _ _ _) 16 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ w _ _ _ _ _ _  _ _ _ _ _ _ _ _) 17 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ w _ _ _ _ _  _ _ _ _ _ _ _ _) 18 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ w _ _ _ _  _ _ _ _ _ _ _ _) 19 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ w _ _ _  _ _ _ _ _ _ _ _) 20 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ w _ _  _ _ _ _ _ _ _ _) 21 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ w _  _ _ _ _ _ _ _ _) 22 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ w  _ _ _ _ _ _ _ _) 23 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  w _ _ _ _ _ _ _) 24 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ w _ _ _ _ _ _) 25 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ w _ _ _ _ _) 26 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ w _ _ _ _) 27 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ w _ _ _) 28 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ w _ _) 29 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ w _) 30 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8Word8 _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ _  _ _ _ _ _ _ _ w) 31 ind = testBit w (fromIntegral ind)

allZeroWord8 :: BitSetWord8Word8
allZeroWord8 = force $ BitSetWord8Word8 0 0 0 0 0 0 0 0  0 0 0 0 0 0 0 0  0 0 0 0 0 0 0 0  0 0 0 0 0 0 0 0

allOneWord8 :: BitSetWord8Word8
allOneWord8 = force $ BitSetWord8Word8 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff
                                       0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff
                                       0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff
                                       0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff

main :: IO ()
main = defaultMain
    [ bgroup "Word64" [ bench "All zero, LSB" $ whnf (memberWord64 allZeroWord64) 0
                      , bench "All zero, 64" $ whnf (memberWord64 allZeroWord64) 64
                      , bench "All zero, 128" $ whnf (memberWord64 allZeroWord64) 128
                      , bench "All zero, MSB" $ whnf (memberWord64 allZeroWord64) 255
                      , bench "All one, LSB" $ whnf (memberWord64 allOneWord64) 0
                      , bench "All one, 64" $ whnf (memberWord64 allOneWord64) 64
                      , bench "All one, 128" $ whnf (memberWord64 allOneWord64) 128
                      , bench "All one, MSB" $ whnf (memberWord64 allOneWord64) 255
                      ]
    , bgroup "ByteString" [ bench "All zero, LSB" $ whnf (memberByteString allZeroByteString) 0
                          , bench "All zero, 64" $ whnf (memberByteString allZeroByteString) 64
                          , bench "All zero, 128" $ whnf (memberByteString allZeroByteString) 128
                          , bench "All zero, MSB" $ whnf (memberByteString allZeroByteString) 255
                          , bench "All one, LSB" $ whnf (memberByteString allOneByteString) 0
                          , bench "All one, 64" $ whnf (memberByteString allOneByteString) 64
                          , bench "All one, 128" $ whnf (memberByteString allOneByteString) 128
                          , bench "All one, MSB" $ whnf (memberByteString allOneByteString) 255
                          ]
    , bgroup "Vector64" [ bench "All zero, LSB" $ whnf (memberVector64 allZeroVector64) 0
                        , bench "All zero, 64" $ whnf (memberVector64 allZeroVector64) 64
                        , bench "All zero, 128" $ whnf (memberVector64 allZeroVector64) 128
                        , bench "All zero, MSB" $ whnf (memberVector64 allZeroVector64) 255
                        , bench "All one, LSB" $ whnf (memberVector64 allOneVector64) 0
                        , bench "All one, 64" $ whnf (memberVector64 allOneVector64) 64
                        , bench "All one, 128" $ whnf (memberVector64 allOneVector64) 128
                        , bench "All one, MSB" $ whnf (memberVector64 allOneVector64) 255
                        ]
    , bgroup "Vector8" [ bench "All zero, LSB" $ whnf (memberVector8 allZeroVector8) 0
                       , bench "All zero, 64" $ whnf (memberVector8 allZeroVector8) 64
                       , bench "All zero, 128" $ whnf (memberVector8 allZeroVector8) 128
                       , bench "All zero, MSB" $ whnf (memberVector8 allZeroVector8) 255
                       , bench "All one, LSB" $ whnf (memberVector8 allOneVector8) 0
                       , bench "All one, 64" $ whnf (memberVector8 allOneVector8) 64
                       , bench "All one, 128" $ whnf (memberVector8 allOneVector8) 128
                       , bench "All one, MSB" $ whnf (memberVector8 allOneVector8) 255
                       ]
    , bgroup "Word8" [ bench "All zero, LSB" $ whnf (memberWord8 allZeroWord8) 0
                     , bench "All zero, 64" $ whnf (memberWord8 allZeroWord8) 64
                     , bench "All zero, 128" $ whnf (memberWord8 allZeroWord8) 128
                     , bench "All zero, MSB" $ whnf (memberWord8 allZeroWord8) 255
                     , bench "All one, LSB" $ whnf (memberWord8 allOneWord8) 0
                     , bench "All one, 64" $ whnf (memberWord8 allOneWord8) 64
                     , bench "All one, 128" $ whnf (memberWord8 allOneWord8) 128
                     , bench "All one, MSB" $ whnf (memberWord8 allOneWord8) 255
                     ]
    ]


src/Data/BitSetWord8.hs
{-|
Module      : Network.BitSetWord8
Copyright   : (c) Naoto Shimazaki 2017
License     : MIT (see the file LICENSE)

Maintainer  : https://github.com/nshimaza
Stability   : experimental

Space efficient set of 'Word8' and some pre-canned sets useful for parsing HTTP related 'ByteString'.
This packaged is intended to provide O(1) membership test on any subset of ASCII and Latin-1 character set
in order to write efficient HTTP related parser.

=== Creating your own set

You can create your own set by 'fromList'.

@
myCharSet :: BitSetWord8
myCharSet = fromList [ 'Y', 'y', 'N', 'n' ]
@

You can create pre-evaluated set using Template Haskell.

@
{-# LANGUAGE TemplateHaskell #-}
import Language.Haskell.TH.Syntax (Lift, lift)

myPreEvaluatedCharSet :: BitSetWord8
myPreEvaluatedCharSet = $(lift myCharSet)
@

=== Example Usage

@
import Data.Attoparsec.ByteString

-- | Parse RFC7230 token.
token :: Parser ByteString
token = takeWhile1 (member rfc7230TChar)
@
-}

{-# LANGUAGE DeriveLift         #-}

module Data.BitSetWord8
    (
    -- * Types
      BitSetWord8
    -- * Useful charset constants in BitSetWord8
    , rfc3986UriReference
    , rfc7230TChar
    , rfc7230QDText
    , rfc7230QuotedPair
    -- * Source Constants
    , rfc3986UriReference'
    , rfc7230QDText'
    , rfc7230QuotedPair'
    , rfc5234Digit'
    , rfc2616UpAlpha'
    , rfc2616LoAlpha'
    , rfc5234Alpha'
    , rfc5234HexDig'
    , rfc5234VChar'
    , rfc5324Wsp'
    , rfc3986SubDelims'
    , rfc3986GenDelims'
    , rfc3986Reserved'
    , rfc3986Unreserved'
    , rfc3986PctEncodedChar'
    , rfc3986PChar'
    , rfc7230TChar'
    , rfc7230ObsText'
    -- * Functions
    , fromList
    , member
    ) where

import Data.BitSetWord8.Internal
import Data.BitSetWord8.CharSets


src/Data/BitSetWord8/CharSets.hs
{-|
Module      : Network.BitSetWord8.CharSets
Copyright   : (c) Naoto Shimazaki 2017
License     : MIT (see the file LICENSE)

Maintainer  : https://github.com/nshimaza
Stability   : experimental

Some pre-canned character sets useful for HTTP related parsing.
All sets provided by this module are evaluated at compile time using Template Haskell.
-}

{-# LANGUAGE TemplateHaskell #-}

module Data.BitSetWord8.CharSets
    (
      rfc3986UriReference
    , rfc7230TChar
    , rfc7230QDText
    , rfc7230QuotedPair
    ) where

import           Language.Haskell.TH.Syntax (Lift, lift)

import           Data.BitSetWord8.Internal

-- | URI-Reference of RFC3986 in BitSetWord8.  Evaluated at compile time.
rfc3986UriReference :: BitSetWord8
rfc3986UriReference = $(lift $ fromList rfc3986UriReference')

-- | tcher of RFC7230 in BitSetWord8.  Evaluated at compile time.
rfc7230TChar :: BitSetWord8
rfc7230TChar = $(lift $ fromList rfc7230TChar')

-- | qdtext of RFC7230 in BitSetWord8.  Evaluated at compile time.
rfc7230QDText :: BitSetWord8
rfc7230QDText = $(lift $ fromList rfc7230QDText')

-- | quoted-string of RFC7230 in BitSetWord8.  Evaluated at compile time.
rfc7230QuotedPair :: BitSetWord8
rfc7230QuotedPair = $(lift $ fromList rfc7230QuotedPair')





src/Data/BitSetWord8/Internal.hs
{-|
Module      : Network.BitSetWord8.Internal
Copyright   : (c) Naoto Shimazaki 2017
License     : MIT (see the file LICENSE)

Maintainer  : https://github.com/nshimaza
Stability   : experimental

Space efficient set of 'Word8' and some pre-canned sets useful for parsing HTTP related 'ByteString'.
This file contains additional useful character sets but they aren't evaluated at compile time.
-}

{-# LANGUAGE DeriveLift    #-}
{-# LANGUAGE MagicHash     #-}
{-# LANGUAGE UnboxedTuples #-}

module Data.BitSetWord8.Internal where

import           Data.Bits                  (setBit, shiftR, testBit)
import           Data.Char                  (chr, ord)
import           Data.List                  (foldl', splitAt)
import           Data.Semigroup             ((<>))
import qualified Data.Set                   as Set (Set, fromList, member)
import           Data.Word                  (Word8, Word64)
import           Instances.TH.Lift
import           Language.Haskell.TH.Syntax (Lift)


-- | Bitwise set of Word8.  Space efficient backend and O(1) membership test.
data BitSetWord8 = BitSetWord8 {-# UNPACK #-} !Word64 {-# UNPACK #-} !Word64
                               {-# UNPACK #-} !Word64 {-# UNPACK #-} !Word64
    deriving (Eq, Lift, Show)

-- | DIGIT of RFC5324 in 'Char' list.
rfc5234Digit' :: [Char]
rfc5234Digit' = ['0'..'9']

-- | UPALPHA of RFC2616 in 'Char' list.
--   Note that RFC2616 has been obsoleted by RFC7230 and RFC7230 doesn't define UPALPHA.
rfc2616UpAlpha' :: [Char]
rfc2616UpAlpha' = [ 'A'..'Z' ]

-- | LOALPHA of RFC2616 in 'Char' list.
--   Note that RFC2616 has been obsoleted by RFC7230 and RFC7230 doesn't define LOALPHA.
rfc2616LoAlpha' :: [Char]
rfc2616LoAlpha' = [ 'a'..'z' ]

-- | ALPHA of RFC5324 in 'Char' list.
rfc5234Alpha' :: [Char]
rfc5234Alpha' = rfc2616UpAlpha' <> rfc2616LoAlpha'

-- | HEXDIG of RFC5324 in 'Char' list.
rfc5234HexDig' :: [Char]
rfc5234HexDig' = rfc5234Digit' <> ['A'..'F']

-- | VCHAR of RFC5324 in 'Char' list.
rfc5234VChar' :: [Char]
rfc5234VChar' = [ '!'..'~']

-- | WSP (aka white space) of RFC5324 in 'Char' list.
rfc5324Wsp' :: [Char]
rfc5324Wsp' = [ '\t', ' ' ]

-- | sub-delim of RFC3986 in 'Char' list.
rfc3986SubDelims' :: [Char]
rfc3986SubDelims' = [ '!', '$', '&', '\'', '(', ')', '*', '+', ',', ';', '=' ]

-- | gen-delim of RFC3986 in 'Char' list.
rfc3986GenDelims' :: [Char]
rfc3986GenDelims' = [ ':', '/', '?', '#', '[', ']', '@']

-- | reserved of RFC3986 in 'Char' list.
rfc3986Reserved' :: [Char]
rfc3986Reserved' = rfc3986GenDelims' <> rfc3986SubDelims'

-- | unreserved of RFC3986 in 'Char' list.
rfc3986Unreserved' :: [Char]
rfc3986Unreserved' = rfc5234Alpha' <> rfc5234Digit' <> [ '-', '.', '_', '~' ]

-- | pct-encoded of RFC3986 in 'Char' list.
rfc3986PctEncodedChar' :: [Char]
rfc3986PctEncodedChar' = ['%'] <> rfc5234HexDig'

-- | pchar of RFC3986 in 'Char' list.
rfc3986PChar' :: [Char]
rfc3986PChar' = rfc3986Unreserved' <> rfc3986PctEncodedChar' <> rfc3986SubDelims' <> [':', '@']

-- | URI-Reference of RFC3986 in 'Char' list.
rfc3986UriReference' :: [Char]
rfc3986UriReference' = rfc3986Reserved' <> rfc3986Unreserved' <> ['%']

-- | tchar of RFC7230 in 'Char' list.
rfc7230TChar' :: [Char]
rfc7230TChar' = [ '!', '#', '$', '%', '&', '\'', '*', '+', '-', '.', '^', '_', '`', '|', '~' ]
    <> rfc5234Digit' <> rfc5234Alpha'

-- | obs-text of RFC7230 in 'Char' list.
rfc7230ObsText' :: [Char]
rfc7230ObsText' = [ chr 0x80 .. chr 0xff]

-- | qdtext of RFC7230 in 'Char' list.
rfc7230QDText' :: [Char]
rfc7230QDText' = rfc5324Wsp' <> [ '!' ] <> [ '#' .. '[' ] <> [ ']' .. '~'] <> rfc7230ObsText'

-- | quoted-pair of RFC7230 in 'Char' list.
rfc7230QuotedPair' :: [Char]
rfc7230QuotedPair' = rfc5324Wsp' <> rfc5234VChar' <> rfc7230ObsText'

-- | O(1).  Return 'True' if given 'Word8' is a member of given 'BitSetWord8'.
member :: BitSetWord8 -> Word8 -> Bool
member bitSet val = doMember bitSet (val `div` 64) (val `mod` 64)
  where
    doMember :: BitSetWord8 -> Word8 -> Word8 -> Bool
    doMember (BitSetWord8 w _ _ _) 0 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8 _ w _ _) 1 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8 _ _ w _) 2 ind = testBit w (fromIntegral ind)
    doMember (BitSetWord8 _ _ _ w) 3 ind = testBit w (fromIntegral ind)

-- | Convert given list of 'Char' into 'Set' of 'Word8'.  Any 'Char' having code point greater than 0xff is ignored.
toWord8Set :: [Char] -> Set.Set Word8
toWord8Set = Set.fromList . map fromIntegral . filter (<= fromIntegral (maxBound :: Word8)) . map ord

-- | Convert 'Set' of Word8 to full filled list of boolean existence flag.
toBoolList :: Set.Set Word8 -> [Bool]
toBoolList wSet = map (\w -> Set.member w wSet) [0..0xff]

-- | Pack 64 of boolean list into single 'Word64' bitwise set.
toWord64 :: [Bool] -> Word64
toWord64 = foldl' (\a e -> let aL = shiftR a 1 in if e == True then setBit aL 63 else aL) 0

-- | Convert full filled boolean list into 4 packed 'Word64' list.
toWord64List :: [Bool] -> [Word64]
toWord64List [] = []
toWord64List bs = let (bs64, rest) = splitAt 64 bs in toWord64 bs64 : toWord64List rest

{- |
    Convert given List of 'Char' into packed bitwise set of Word64.
    Any 'Char' having code point greater than 0xff is ignored.
-}
fromList :: [Char] -> BitSetWord8
fromList cs = BitSetWord8 w0 w1 w2 w3
  where
    (w0:w1:w2:w3:_) = toWord64List . toBoolList . toWord8Set $ cs
